cmdstanpy/model.py

```diff
@@ -601,20 +606,56 @@ def sample(
                             'Chain_id must be a non-negative integer value,'
                             ' found {}.'.format(chain_id)
                         )
-
-        cores_avail = cpu_count()
-        if cores is None:
-            cores = max(min(cores_avail - 2, chains), 1)
-        if cores < 1:
+        if parallel_chains is None or parallel_chains > chains:
+            parallel_chains = max(min(cpu_count(), chains), 1)
+        if parallel_chains > cpu_count():
+            self._logger.warning(
+                'Requested %u parallel chains, but only %u CPUs,'
+                ' can only run %u chains in parallel.',
+                threads_per_chain, cpu_count(), cpu_count()
+            )
+            parallel_chains = max(min(cpu_count(), chains), 1)
+        if parallel_chains < 1:
+            raise ValueError(
+                'Argument parallel_chains must be a positive integer value, '
+                'found {}.'.format(parallel_chains)
+            )
+        if threads_per_chain is None:
+            threads_per_chain = 1
+        if threads_per_chain < 1:
             raise ValueError(
-                'Argument cores must be a positive integer value, '
-                'found {}.'.format(cores)
+                'Argument threads_per_chain must be a positive integer value, '
+                'found {}.'.format(threads_per_chain)
             )
-        if cores > cores_avail:
+        os.environ['STAN_NUM_THREADS'] = str(threads_per_chain)
+        if threads_per_chain > cpu_count():
             self._logger.warning(
-                'Requested %u cores, only %u available.', cores, cpu_count()
+                'Requested threads_per_chain: %u > available CPUs: %u,'
+                ' will oversubcribe CPUs',
+                threads_per_chain,
+                cpu_count(),
             )
-            cores = cores_avail
+            if parallel_chains > 1:
+                parallel_chains = 1
+                self._logger.warning(", won't run chains in parallel")
+        if parallel_chains > 1:
```

Comment:
how much does this affect to runtime?

Reply:
got rid of this code altogether - we don't try to override user requests - will run requested number of parallel chains with requested number of threads.